### PR TITLE
production trace stream path, part 1

### PR DESCRIPTION
**Pull request by Jim3Things**
This change sets up most of the critical infrastructure for creating the trace sink process - which then will be used by the UI for the trace display.

- The production exporter has been added, 
- The stdout exporter has been redone as an IO writer exporter and configuration added to specify the trace file for each service
- The trace init code now accepts multiple exporters.  IO writer + production is the common form
- All exporters now automatically store and forward trace entries based on having a complete connection path.  This handles the chicken and egg issues around config tracing, and handles failure and restart of the trace sink process.

Finally, a few minor lint issues were addressed along the way.

### CODE REVIEW

File: internal/tracing/exporters/production/exporter.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        return
+    }
+
+    ms := reconnectBackoff * connectCount
+    if ms > maxBackoff {
+        ms = maxBackoff
+    }
+
+    state = awaitingConnection
+    _ = time.AfterFunc(time.Duration(ms) * time.Millisecond, func() {
```
Comment (MikeCJohnsonUk) [inline internal/tracing/exporters/production/exporter.go:219]: I presume "time.Duration(ms) * time.Millisecond" is the same as "time.Duration(ms * time.Millisecond)"? Looking at the docks it should be...

File: internal/tracing/exporters/production/exporter.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        mutex.Lock()
+        defer mutex.Unlock()
+
+        if state != awaitingConnection {
+            return
+        }
```
Comment (MikeCJohnsonUk) [inline internal/tracing/exporters/production/exporter.go:225]: Not sure why you need this as the same test is already in attemptConnection()

File: cmd/sim_supportd/main.go [hunk 3]
```diff
@@ -28,7 +29,7 @@ func main() {
 		os.Exit(0)
 	}
 
-	setup.Init(exporters.StdOut)
+	setup.Init(exporters.IoWriter /*, exporters.LocalProduction*/)
```
[the remark this answers is not in the document]
Comment (Jim3Things) [reply cmd/sim_supportd/main.go:32]: Yes.  It was a reminder to myself to use the local production variant in this case, once it was ready to use.

File: internal/tracing/exporters/exporters.go [hunk 4]
```diff
@@ -1,7 +1,19 @@
...
 	UnitTest
+
+	// Production forwards the trace calls to a sink that another caller can
+	// retrieve them from for display on the UI
 	Production
+
+	// LocalProduction is a shortcut used by the process that hosts the
+	// production trace sink to avoid appends to the trace log recursively
+	// expanding forever - and to still be able to trace those calls.
+	LocalProduction
```
[the remark this answers is not in the document]
Comment (Jim3Things) [reply internal/tracing/exporters/exporters.go:18]: Always open to naming.  In this case the exporter only works if it is in the same process as the trace sink, which is why I was thinking 'local'.  Happy to convert to 'internal' if that would be better.

File: internal/tracing/exporters/exporters.go [hunk 5]
```diff
@@ -1,7 +1,19 @@
 package exporters
 
 const (
-	StdOut = iota
+	// IoWriter formats and writes the trace calls to a specified trace file
+	IoWriter = iota
```
[the remark this answers is not in the document]
Comment (Jim3Things) [reply internal/tracing/exporters/exporters.go:5]: Yeah, I started with FileWriter, but in reality it is provide an io.Writer, hence the name...  And it gets that in order to make handling stdout as a choice straightforward...

File: internal/tracing/setup/config.go [hunk 6]
```diff
@@ -1,36 +1,67 @@
...
+// SetFileWriter sets up the IO writer for the trace exporter that outputs to
+// a file.  It defaults to stdout if no file name is specified.
+func SetFileWriter(name string) error {
+	if name == "" || strings.EqualFold(name, "stdout") {
+		// If no trace file specified, use stdout
+		return io_writer.SetLogFileWriter(os.Stdout)
+	} else {
+		writer, err := os.OpenFile(
+			name,
+			os.O_APPEND | os.O_CREATE | os.O_WRONLY,
```
[the remark this answers is not in the document]
Comment (Jim3Things) [reply internal/tracing/setup/config.go:51]: Not sure.  The choices were based on what I'd seen others do.  And we don't read the file, so saying we don't seemed at least not wrong.

File: internal/tracing/exporters/production/exporter.go (hunk 1 above)
Comment (Jim3Things) [reply internal/tracing/exporters/production/exporter.go:219]: Should be.

File: cmd/sim_supportd/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jim3Things) [reply cmd/sim_supportd/main.go:32]: I've moved it to a TODO note ahead of the line

File: internal/config/settings.go [hunk 7]
```diff
@@ -141,17 +149,22 @@ func newGlobalConfig() *GlobalConfig {
...
+			},
+			TraceFile: InventoryDefaultTraceFile,
+		},
 		SimSupport: SimSupportType{
 			EP: Endpoint{
 				Hostname: DefaultHost,
 				Port:     SimSupportDefaultPort,
 			},
+			TraceFile: SimSupportDefaultTraceFile,
 			StepperPolicy: DefaultStepperPolicy},
```
[the remark this answers is not in the document]
Comment (Jim3Things) [reply internal/config/settings.go:168]: done.

File: internal/tracing/exporters/init.go [hunk 8]
```diff
@@ -4,26 +4,37 @@ import (
...
 	switch exportType {
-	case StdOut:
-		exporter, err = stdout.NewExporter(stdout.Options{PrettyPrint: true})
-		if err != nil {
+	case IoWriter:
+		if exporter, err := io_writer.NewExporter(); err != nil {
 			log.Fatal(err)
+		} else {
+			return exporter, nil
 		}
```
[the remark this answers is not in the document]
Comment (Jim3Things) [reply internal/tracing/exporters/init.go:22]: done.

File: internal/tracing/exporters/unit_test/exporter.go [hunk 9]
```diff
@@ -15,57 +15,65 @@ import (
...
 	defer mutex.Unlock()
 
 	if testContext != nil {
-		flushSaved()
-		processOneEntry(common.ExtractEntry(ctx, data), false)
+		flushSaved(ctx)
+		processOneEntry(entry, false)
 	} else {
-		savedEntries = append(savedEntries, common.ExtractEntry(ctx, data))
+		_= queue.Defer(entry)
```
[the remark this answers is not in the document]
Comment (Jim3Things) [reply internal/tracing/exporters/unit_test/exporter.go:67]: done

File: internal/tracing/exporters/production/exporter.go (hunk 2 above)
Comment (Jim3Things) [reply internal/tracing/exporters/production/exporter.go:225]: Mostly paranoia.  I should probably put a log.Fatal there.  This line occurs in the action function, which is asynchronous wrt the rest of the path, and is a belt-and-braces check that the state machine has not moved on.

File: internal/tracing/exporters/production/exporter.go (hunk 1 above)
Comment (Jim3Things) [reply internal/tracing/exporters/production/exporter.go:219]: .. but doesn't appear to be.  The latter gets a mismatched type error (int vs. Duration)

File: internal/config/settings.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply internal/config/settings.go]: Thank you!

**Pull request merged**